### PR TITLE
New dustc compiler

### PR DESCRIPTION
**Pull request by sethkinast**
This depends on #536 so it's a branch off that branch. That means that #536 changes appear in this PR; however there are only changes to `dustc` in this branch.

I also added a unit test suite to cover `dustc`, since we previously had no test coverage on the command-line compiler.

### CODE REVIEW

**Comment by prashn64** [comment]
lgtm, let's get one more set of eyes on it.

@jimmyhchan @smfoote @rragan

File: bin/dustc [hunk 1]
```diff
@@ -1,82 +1,120 @@
...
 
-var ARG_REGEX = /^--?([a-z][0-9a-z-]*)(?:=([^\s]+))?$/i,
-    args = process.argv.slice(1),
-    name = null;
+cli.setApp('dustc', require('../package.json').version);
 
-args = args.filter(function(arg) {
-    var match = arg.match(ARG_REGEX),
-        val;
+cli.setUsage(cli.app + ' [options] [path1 [path2 path3...]]\n\n' +
```
Comment (jimmyhchan) [inline bin/dustc]: Is this an API change. No back compat?

File: bin/dustc [hunk 2]
```diff
@@ -1,82 +1,120 @@
...
     });
+});
+
+/*** helper functions ***/
 
-    process.stdin.on('end', function() {
-        parseDustFile(false, buffer);
+function read(filename, cb) {
+    var data = '',
+        file = cli.native.fs.createReadStream(filename);
```
Comment (jimmyhchan) [inline bin/dustc]: nit: may as well put fs at the top of the file with the others.

File: bin/dustc [hunk 3]
```diff
@@ -1,82 +1,120 @@
...
+ * Handle a list of paths passed as args
+ * Each path gets globbed in case the OS doesn't support it
+ */
+cli.args
+.map(function(arg) { return cli.glob.sync(arg); })
+.reduce(function(a, b) { return a.concat(b); }, [])
+.forEach(function(inputFile, index, filesToProcess) {
+    read(inputFile, function(err, data) {
+        if (err) {
+            return;
```
Comment (jimmyhchan) [inline bin/dustc]: please log or fatal here as well.

File: bin/dustc (hunk 2 above)
Comment (jimmyhchan) [inline bin/dustc]: You should patch sindresorhus/gulp-dust to allow stream support

**Comment by jimmyhchan** [comment]
:+1:   looks good. small comments. Good stuff. Thanks!

File: bin/dustc (hunk 3 above)
Comment (sethkinast) [inline bin/dustc]: Makes sense, thanks

File: bin/dustc (hunk 1 above)
Comment (sethkinast) [inline bin/dustc]: Yep, new compiler. Had to kill the old syntax to allow it to work like a shell script.

`dustc input output` doesn't work anymore.

That's why I waited for a new minor version.

File: bin/dustc (hunk 2 above)
Comment (sethkinast) [inline bin/dustc]: Good catch.

**Pull request merged**